# Homework 2 (Due 6:29pm PST April 6th, 2021): Word Vectorization, Regex Practice, and Similarity

You may work with **one other person on this assignment**. You may also work independently if you prefer.

If you just want to be assigned someone to work with, message me on Slack and I will assign you a partner to work with.

A. Using the **McDonalds Yelp Review CSV file**, **process the reviews**.
This means you should think briefly about:
* what stopwords to remove (should you add any custom stopwords to the set? Remove any stopwords?)
* what regex cleaning you may need to perform (for example, are there different ways of saying `hamburger` that you need to account for?)
* stemming/lemmatization (explain in your notebook why you used stemming versus lemmatization). 

Next, **count-vectorize the dataset**. Use the **`sklearn.feature_extraction.text.CountVectorizer`** examples from `Linear Algebra, Distance and Similarity (Completed).ipynb` and `Text Preprocessing Techniques (Completed).ipynb` (read the last section, `Vectorization Techniques`).

I do not want redundant features - for instance, I do not want `hamburgers` and `hamburger` to be two distinct columns in your document-term matrix. Therefore, I'll be taking a look to make sure you've properly performed your cleaning, stopword removal, etc. to reduce the number of dimensions in your dataset. 


In [1]:
text_file = open('mcdonalds-yelp-negative-reviews.csv','r',encoding='latin1')
lines = text_file.readlines()

In [2]:
import re

In [3]:
regex = re.compile(r'[^\w\s\']')

In [4]:
regex_digit = re.compile(r'\b\d+\b')

In [5]:
lst = []
for line in lines:
    new_line = regex.sub(' ',line)
    new_line = regex_digit.sub('',new_line)
    lst.append(new_line)

In [6]:
#d

In [7]:
d = {}
for line in lst:
    for word in line.split():
        d[word] = d.get(word,0) + 1

# dict(sorted(d.items(), key=lambda x: x[1], reverse = True))     

In [8]:
def word_count(lst):
    d = {}
    for line in lst:
        for word in line.split():
            d[word] = d.get(word,0) + 1
    return d

In [9]:
from nltk.corpus import stopwords

In [10]:
stopword_set = set(stopwords.words('english'))

In [11]:
added = ['mcdonald\'s','mcdonalds','mcdonald','mcds','mcd\'s','i\'m','i\'ve',\
         'got','get','order','ordered','grab','grabbed','go','went','think','buy','bought',\
         'really','also','even','always','sometimes','right','star','pm','am','going',\
         'aaaaaaaahhhhhhhhhhh','','yet','oh','one','two']

stopword_set.update(added)
#stopword_set

In [12]:
cleaned = []
for line in lst:
    newline = []
    for word in line.split():
        if word.lower() in stopword_set:
            continue
        else:
            newline.append(word.lower())
    review = ' '.join(newline)
    cleaned.append(review)


In [13]:
count_dic = word_count(cleaned)
# dict(sorted(count_dic.items(), key=lambda x: x[1], reverse = True))

In [14]:
test = cleaned[1]


In [15]:
from nltk.stem import PorterStemmer
from nltk.tokenize import word_tokenize

In [16]:
stemmer = PorterStemmer()
stemmed = []

for line in cleaned:
    new_line = []
    for word in word_tokenize(line):
        new_word = stemmer.stem(word)
        new_line.append(new_word)
    stemmed_review = ' '.join(new_line)
    stemmed.append(stemmed_review )
# stemmed

In [17]:
from sklearn.feature_extraction.text import CountVectorizer
vectorizer = CountVectorizer()

In [18]:
X = vectorizer.fit_transform(stemmed)

In [19]:
X.shape

(1526, 5929)

In [20]:
X = X.toarray()

In [21]:
import pandas as pd

In [22]:
df = pd.DataFrame(X, columns = vectorizer.get_feature_names())
df.head()

00am  00mi  00pm  03pm  04am  05i  07am  09chocol  0food  0overal  ...  \
0     0     0     0     0     0    0     0         0      0        0  ...   
1     0     0     0     0     0    0     0         0      0        0  ...   
2     0     0     0     0     0    0     0         0      0        0  ...   
3     0     0     0     0     0    0     0         0      0        0  ...   
4     0     0     0     0     0    0     0         0      0        0  ...   

   zak  zax  zee  zeke  zero  zesti  zip  zombi  zoom  î_  
0    0    0    0     0     0      0    0      0     0   0  
1    0    0    0     0     0      0    0      0     0   0  
2    0    0    0     0     0      0    0      0     0   0  
3    0    0    0     0     0      0    0      0     0   0  
4    0    0    0     0     0      0    0      0     0   0  

[5 rows x 5929 columns]

In [23]:
# for i in range(5900):
#     print(df.columns[i])

### Summary for question A:

In this question, I first eliminate all the punctuations and long number string (e.g. strings like 67897652). I then import the common stop words from the library. I also added several stopwords such as McDonalds, order, get, etc. After cleaning the dataset, I ulterlized the method stemming in order to further reduce the dimensionalty. Finally, the count_vectorize method was used to transform the file to a list of vectors. At the end, the shape of each vector is 5938.

B. **Stopwords, Stemming, Lemmatization Practice**

Using the `tale-of-two-cities.txt` file from Week 1:
* Count-vectorize the corpus. Treat each sentence as a document.

How many features (dimensions) do you get when you:
* Perform **stemming** and then **count-vectorization**.
* Perform **lemmatization** and then **count-vectorization**.
* Perform **lemmatization**, remove **stopwords**, **remove punctuation**, and then perform **count-vectorization**?

In [24]:
path = '/Users/xinmengqiao/Desktop/DSO560/dso-560-nlp-text-analytics-SPRING-2021/Week 1/'

text_file1 = open(path+'tale-of-two-cities.txt','r')

In [25]:
lines = text_file1.readlines()

In [26]:
# import nltk
# nltk.download('punkt')

In [27]:
from nltk.tokenize import sent_tokenize

In [28]:
all_text = ' '.join(lines)
new_lst = sent_tokenize(all_text)

In [29]:
stemmer = PorterStemmer()
stemmed = []

for line in new_lst:
    new_line = []
    for word in word_tokenize(line):
        new_word = stemmer.stem(word)
        new_line.append(new_word)
    stemmed_sen = ' '.join(new_line)
    stemmed.append(stemmed_sen )
# stemmed

In [30]:
vectorizer_stem = CountVectorizer(token_pattern=r"(?u)\b\w\w+\b|!|\?|\"|\'")
X = vectorizer_stem.fit_transform(stemmed)
X = X.toarray()
X.shape

(7764, 6678)

In [31]:
from nltk.stem import WordNetLemmatizer

In [32]:
# import nltk
# nltk.download('wordnet')

In [33]:
lemma = WordNetLemmatizer()
lemmatized = []

for line in new_lst:
    new_line = []
    for word in word_tokenize(line):
        new_word = lemma.lemmatize(word)
        new_line.append(new_word)
    sen = ' '.join(new_line)
    lemmatized.append(sen)
# lemmatized

In [34]:
vectorizer_stem = CountVectorizer(token_pattern=r"(?u)\b\w\w+\b|!|\?|\"|\'")
X = vectorizer_stem.fit_transform(lemmatized)
X = X.toarray()
X.shape

(7764, 8919)

In [35]:
stopword_set = set(stopwords.words('english'))
cleanned = []
for line in lemmatized:
    new_line = []
    for word in line.split():
        if word.lower() in stopword_set:
            continue
        new_line.append(word)
    new_sen = ' '.join(new_line)
    cleanned.append(new_sen)
# cleanned

In [36]:
vectorizer_stem = CountVectorizer()
X = vectorizer_stem.fit_transform(cleanned)
X = X.toarray()
X.shape

(7764, 8857)

### Summary for PartB
In this part, I performed count_vectorize on the text file -- tale of two cities. Only performing stemming and count-vectorize gives me a dimension of 6678. Performing lemmatization and count-vectorizing gives a dimension of 9819. After removing stopwords and punctuations, the count-vectorizer will give a final dimension of 8857.

## Classwork For Lecture 2 (Due 6:29pm PST March 30th, 2021): Word Vectorization, Regex Practice, and Similarity

#### Pick A or B

A. Answer all the exercise questions in the **Text Preprocessing** notebook.

B. Answer the below questions about text encoding and word count distributions:

1. Which of the encodings below will be able to encode this text: 사업.
* `ascii`
* `latin1`
* `utf-8`
* `utf-32`
* `extended ascii`

2. **True or False**: the word dog will have the same binary representation regardless of whether it is ASCII, latin1, or utf8. If False, explain why it is false.


3. According to the Zipf Law approximation, approximately what frequency (express it has a percent) would the 3rd most popular word in a generic piece of text appear?


4. **True or False**: what is considered a stopword changes depending on the business context and dataset you are working with. If true, provide an example. If false, explain why it is false.

4, 5
T
0.033
t
